File: src/grid/finder.ts
import { Point, FindPathConfig } from "../types/main.ts";
import { Grid } from "./grid.ts";
import { OpenList } from "./openList.ts";

const NOT_REACHED_COST: number = 999999;
const MAX_JUMP_HEIGHT: number = 5;

class PathNode {
  public from: PathNode | null;
  public point: Point;
  public cost: number;
  public heuristicValue: number;

  constructor(
    point: Point,
    from: PathNode | null,
    cost: number,
    heuristicValue: number,
  ) {
    this.from = from;
    this.point = point;
    this.cost = cost;
    this.heuristicValue = heuristicValue;
  }

  toString(): string {
    return `(${this.point.x}, ${this.point.y}, cost: ${this.cost}, heuristic: ${this.heuristicValue})`;
  }
}

export const findPath = (
  startPoint: Point,
  endPoint: Point,
  grid: Grid,
  config: FindPathConfig,
): Point[] => {
  const diagonalCostMultiplier = config.diagonalCostMultiplier ?? 1;
  const orthogonalCostMultiplier = config.orthogonalCostMultiplier ?? 1;
  const maxJumpCost = config.maxJumpCost ?? 5;
  const maxIterations = config.maxIterations ?? 99999;
  const jumpBlockedDiagonals = config.jumpBlockedDiagonals ?? false;

  const index = (point: Point): number => {
    return point.y * grid.height + point.x;
  };

  const getMoveCostAt = (src: Point, dst: Point): number | null => {
    if (!grid.inBounds(src) || !grid.inBounds(dst)) return null;

    // Difference in height
    const srcHeight = grid.getHeightAt(src) ?? NOT_REACHED_COST;
    const dstHeight = grid.getHeightAt(dst) ?? NOT_REACHED_COST;

    // Max jump
    if (Math.abs(srcHeight - dstHeight) > MAX_JUMP_HEIGHT) return null;

    return 1;
  };

  // Orthogonal jumps from JumpPoint
  const addOrthogonalJumps = (
    prevNode: PathNode,
    src: Point,
    srcCost: number,
    dirX: number,
    dirY: number,
  ) => {
    let jumpDistance = 1;
    let accumulatedCost = 0;
    let prevPoint = src;

    while (true) {
      const target: Point = {
        x: src.x + dirX * jumpDistance,
        y: src.y + dirY * jumpDistance,
      };

      if (!grid.isWalkable(target)) {
        break;
      }

      const moveCost = getMoveCostAt(prevPoint, target);
      if (moveCost === null) {
        break;
      }

      accumulatedCost += moveCost * orthogonalCostMultiplier;
      const targetIndex = index(target);
      const totalCost = srcCost + accumulatedCost;

      if (totalCost < visited[targetIndex]) {
        visited[targetIndex] = totalCost;
        queue.push(
          new PathNode(target, prevNode, totalCost, heuristic(target)),
        );
      }
      prevPoint = target;
      jumpDistance++;

      if (accumulatedCost > maxJumpCost) {
        break;
      }
    }
  };

  // Diagonal movements with filter for adjacent walkable tiles in the same floor height
  const addDiagonal = (
    prevNode: PathNode,
    src: Point,
    srcCost: number,
    dirX: number,
    dirY: number,
  ) => {
    const target: Point = { x: src.x + dirX, y: src.y + dirY };
    const moveCost =
      srcCost +
      (getMoveCostAt(src, target) ?? NOT_REACHED_COST) * diagonalCostMultiplier;
    const targetHeight = grid.getHeightAt(target);
    const aux1: Point = { x: src.x, y: src.y + dirY };
    const aux2: Point = { x: src.x + dirX, y: src.y };
    const targetIndex = index(target);

    const canJumpDiagonals =
      jumpBlockedDiagonals ||
      (grid.isWalkable(aux1) &&
        grid.isWalkable(aux2) &&
        targetHeight == grid.getHeightAt(aux1) &&
        targetHeight == grid.getHeightAt(aux2));

    if (
      grid.isWalkable(target) &&
      canJumpDiagonals &&
      moveCost < visited[targetIndex]
    ) {
      visited[targetIndex] = moveCost;
      queue.push(new PathNode(target, prevNode, moveCost, heuristic(target)));
    }
  };

  if (!grid.isWalkable(startPoint) || !grid.isWalkable(endPoint)) {
    return [];
  }

  const visited = new Float32Array(grid.width * grid.height);
  visited.fill(NOT_REACHED_COST);
  config.travelCosts = visited;

  // Distance to the end point, from A*
  const travelHeuristic = new Float32Array(grid.width * grid.height);
  travelHeuristic.fill(NOT_REACHED_COST);

  grid.walkMatrix((x, y) => {
    travelHeuristic[y * grid.height + x] = grid.distance({ x, y }, endPoint);
  });
  config.travelHeuristic = travelHeuristic;

  const heuristic = (a: Point): number => {
    return travelHeuristic[index(a)];
  };

  const comparator = (a: PathNode, b: PathNode): number => {
    return a.cost + a.heuristicValue - (b.cost + b.heuristicValue);
  };

  const queue: OpenList<PathNode> = new OpenList(comparator);
  let iterations = 0;

  visited[index(startPoint)] = 0;
  queue.push(new PathNode(startPoint, null, 0, heuristic(startPoint)));

  while (!queue.isEmpty()) {
    if (iterations > maxIterations) {
      return [];
    }
    iterations++;

    const node = queue.pop();
    const point = node.point;

    // If we reached the end point, return the path to it
    if (point.x === endPoint.x && point.y === endPoint.y) {
      return pathFromNode(node);
    }

    addOrthogonalJumps(node, point, node.cost, 0, -1);
    addOrthogonalJumps(node, point, node.cost, 0, 1);
    addOrthogonalJumps(node, point, node.cost, -1, 0);
    addOrthogonalJumps(node, point, node.cost, 1, 0);

    addDiagonal(node, point, node.cost, 1, 1);
    addDiagonal(node, point, node.cost, -1, 1);
    addDiagonal(node, point, node.cost, 1, -1);
    addDiagonal(node, point, node.cost, -1, -1);
  }

  // No path found
  return [];
};

const pathFromNode = (lastNode: PathNode): Point[] => {
  const path: Point[] = [];
  let node: PathNode | null = lastNode;
  while (node) {
    path.push(node.point);
    node = node.from;
  }
  return path.reverse();
};

File: src/finder.rs
use crate::find_path_config::FindPathConfig;
use crate::grid::Grid;
use crate::open_list::{Compare, OpenList};
use crate::path_node::PathNode;
use crate::point::Point;
use std::f32;

#[derive(Debug)]
pub struct PathNodeComparator;

impl Compare<PathNode> for PathNodeComparator {
    fn compare(&self, a: &PathNode, b: &PathNode) -> f32 {
        (a.cost + a.heuristic_value) - (b.cost + b.heuristic_value)
    }
}

pub struct Finder {
    not_reached_cost: f32,
    visited: Vec<f32>,
    travel_heuristic: Vec<f32>,
    queue: OpenList<PathNode>,

    pub start_point: Point,
    pub end_point: Point,
    pub grid: Grid,
    pub config: FindPathConfig,
}

impl Finder {
    pub fn new(
        start_point: Point,
        end_point: Point,
        grid: Grid,
        config: Option<FindPathConfig>,
    ) -> Self {
        let not_reached_cost = 999999.0;

        let mut visited = vec![0.0; grid.width * grid.height];
        visited.fill(not_reached_cost);

        let mut travel_heuristic = vec![0.0; grid.width * grid.height];
        travel_heuristic.fill(not_reached_cost);

        Finder {
            not_reached_cost,
            visited,
            travel_heuristic,
            queue: OpenList::new(Box::new(PathNodeComparator)),

            start_point,
            end_point,
            grid,
            config: config.unwrap_or(FindPathConfig::new(None, None, None, None, None)),
        }
    }

    fn index(&self, point: &Point) -> usize {
        (point.y * (self.grid.height as isize) + point.x) as usize
    }

    fn get_move_cost_at(&self, src: &Point, dst: &Point) -> Option<f32> {
        if !self.grid.in_bounds(&src) || !self.grid.in_bounds(&dst) {
            return None;
        }
        let src_height = self
            .grid
            .get_height_at(&src)
            .unwrap_or(self.not_reached_cost);
        let dst_height = self
            .grid
            .get_height_at(&dst)
            .unwrap_or(self.not_reached_cost);

        println!(
            "{} {}",
            (src_height - dst_height).abs(),
            self.config.max_jump_cost
        );
        if (src_height - dst_height).abs() > self.config.max_jump_cost {
            return None;
        }

        Some(1.0)
    }

    fn heuristic(&self, point: &Point) -> f32 {
        self.travel_heuristic[self.index(&point)]
    }

    fn add_orthogonal_jumps(&mut self, prev_node: &PathNode, dir_x: isize, dir_y: isize) {
        let mut jump_distance = 1;
        let mut accumulated_cost = 0f32;
        let mut prev_point = prev_node.point.clone();

        loop {
            let target: Point = Point::new(
                prev_node.point.x + dir_x * jump_distance,
                prev_node.point.y + dir_y * jump_distance,
            );
            if !self.grid.is_walkable(&target) {
                return;
            }
            let move_cost = match self.get_move_cost_at(&prev_point, &target) {
                Some(cost) => cost,
                None => return,
            };

            accumulated_cost += move_cost * self.config.orthogonal_cost_multiplier;
            let target_index = self.index(&target);
            let total_cost = prev_node.cost + accumulated_cost;

            if total_cost < self.visited[target_index] {
                self.visited[target_index] = total_cost;
                self.queue.push(PathNode::new(
                    self.start_point.clone(),
                    None,
                    0f32,
                    self.heuristic(&target),
                ))
            }
            prev_point = target;
            jump_distance = jump_distance + 1;

            if accumulated_cost > self.config.max_jump_cost {
                return;
            }
        }
    }

    fn add_diagonal(&mut self, prev_node: &PathNode, dir_x: isize, dir_y: isize) {
        let target: Point = Point {
            x: prev_node.point.x + dir_x,
            y: prev_node.point.y + dir_y,
        };
        let move_cost = prev_node.cost
            + (self
                .get_move_cost_at(&prev_node.point, &target)
                .unwrap_or(self.not_reached_cost)) as f32
                * self.config.diagonal_cost_multiplier;
        let target_height = self.grid.get_height_at(&target);
        let aux1: Point = Point {
            x: prev_node.point.x,
            y: prev_node.point.y + dir_y,
        };
        let aux2: Point = Point {
            x: prev_node.point.x + dir_x,
            y: prev_node.point.y,
        };
        let target_index = self.index(&target);

        let can_jump_diagonals = self.config.jump_blocked_diagonals
            || (self.grid.is_walkable(&aux1)
                && self.grid.is_walkable(&aux2)
                && target_height == self.grid.get_height_at(&aux1)
                && target_height == self.grid.get_height_at(&aux2));

        if self.grid.is_walkable(&target)
            && can_jump_diagonals
            && move_cost < self.visited[target_index]
        {
            self.visited[target_index] = move_cost;
            self.queue.push(PathNode::new(
                target.clone(),
                Some(Box::new(prev_node.clone())),
                move_cost,
                self.heuristic(&target),
            ))
        }
    }

    fn get_path_from_node(&self, last_node: PathNode) -> Vec<Point> {
        let mut path: Vec<Point> = Vec::new();
        let mut node: Option<Box<PathNode>> = Some(Box::new(last_node));
        loop {
            match node {
                None => {
                    break;
                }
                Some(current_node) => {
                    path.push(current_node.point);
                    node = current_node.from
                }
            }
        }

        path.reverse();
        path
    }

    pub fn find(&mut self) -> Vec<Point> {
        let empty = Vec::new();

        if !self.grid.is_walkable(&self.start_point) || !self.grid.is_walkable(&self.end_point) {
            return empty;
        }

        self.grid.walk_matrix(|x, y, _cost| {
            self.travel_heuristic[y * self.grid.height + x] = self
                .grid
                .distance(&Point::new(x as isize, y as isize), &self.end_point);
        });

        let start_index = self.index(&self.start_point.clone());
        self.visited[start_index] = 0f32;
        self.queue.push(PathNode::new(
            self.start_point.clone(),
            None,
            0f32,
            self.heuristic(&self.start_point),
        ));

        let mut iterations = 0;

        loop {
            if self.queue.is_empty() {
                return empty;
            }
            if iterations > self.config.max_iterations {
                return empty;
            }
            iterations += 1;
            let node = self.queue.pop().unwrap();

            if node.point.x == self.end_point.x && node.point.y == self.end_point.y {
                return self.get_path_from_node(node);
            }

            self.add_orthogonal_jumps(&node, 0, -1);
            self.add_orthogonal_jumps(&node, 0, 1);
            self.add_orthogonal_jumps(&node, -1, 0);
            self.add_orthogonal_jumps(&node, 1, 0);

            self.add_diagonal(&node, 1, 1);
            self.add_diagonal(&node, -1, 1);
            self.add_diagonal(&node, 1, -1);
            self.add_diagonal(&node, -1, -1);
        }
    }
}
